### ISSUE TITLE
Tracking issue: metadata filtering

### ISSUE THREAD

**Issue by asg017** [asg017/sqlite-vec#26]
`sqlite-vec` doesn't have good metadata filtering as of `v0.1.0`. Only vector columns can be declared in the `vec0` constructor. You can do pre-filtering with `vec_column IN (...)` queries, but that's slow and inconvenient. 

I'm thinking:

```sql
create virtual table vec_movies(
  movie_id text primary key,
  genre text,
  release_date date,
  rating text,
  is_3d boolean,
  synopsis_embedding float[768]
);
```

`genre`, `release_date`, `rating`, and `is_3d` would all be "metadata" columns. You could do queries like:

```sql
select
  rowid,
  distance
from vec_movies
where synopsis_embedding match embed('comedic american summer camp')
  and k = 20
  and is_3d
  and release_date between '2010-01-01' and '2015-12-31'
  and rating = 'PG';
```

We could capture all the `WHERE` clauses to ensure that the top 20 returned vectors match that criteria.

A few open questions:

## How do we store metadata values?

We could store in OLTP-fashion with the `_rowids` shadow tables, but that may be slow. We could store in column-oriented fashion to match the vector column formats, but unsure how much faster that would be. 

## How would this work with ANN indexes?

🤷 

## What datatypes to support?

Ideally everything, ideally `STRICT`. But if we do column-oriented we'd need a strict subset. like:

- `TEXT`
- `INT`
- `DOUBLE`
- `BLOB`
- `BOOLEAN`
- `DATE`/`DATETIME`

Maybe we could do dictionary encoding for text values? maybe that's a column option, like `genre text encoding=dictionary` or something. Maybe `ENUM`s? `NULL`/`NOT NULL`?

**Comment by forrestbao** [comment]
I really wish this feature can be available soon.

**Comment by ajram23** [comment]
@asg017 +1

**Comment by lojik-ng** [comment]
+1

**Comment by charnould** [comment]
@asg017 First, thanks for making `sqlite-vec`! `SQLite` is an amazing piece of software, and your extension a bless! `pre-filtering` is super important/useful for any RAG app. Do you have any ETA for this? Thanks again.

**Comment by asg017** [comment]
@charnould the PR is #124 but it needs tests and docs. Aiming for public release in 3 weeks (~Nov 19th) but hoping to get a beta out before then that you can try!

**Comment by ajram23** [comment]
@asg017 happy to beta test when you release it!

**Comment by asg017** [comment]
As of [`v0.1.6`](https://github.com/asg017/sqlite-vec/releases/tag/v0.1.6), metadata columns are now supported in `sqlite-vec`! See the [announcement blog post](https://alexgarcia.xyz/blog/2024/sqlite-vec-metadata-release/index.html) for more info, and reference the [`sqlite-vec` metadata documentation](https://alexgarcia.xyz/sqlite-vec/features/vec0.html#metadata-in-vec0-virtual-tables) for more details.


tldr you have a choice between storing metdata in `vec0` virtual tables with either metadata column, partition keys, and auxiliary columns, all with their own benefits/drawbacks. Reference the blog post or docs for a more detailed look.

Closing this issue now, feel free to file more bugs/issues if needed!
